src/Vehicle/CompInfoGeneral.cc

```diff
@@ -0,0 +1,93 @@
+/****************************************************************************
+ *
+ * (c) 2009-2020 QGROUNDCONTROL PROJECT <http://www.qgroundcontrol.org>
+ *
+ * QGroundControl is licensed according to the terms in the file
+ * COPYING.md in the root of the source code directory.
+ *
+ ****************************************************************************/
+
+#include "CompInfoGeneral.h"
+
+#include "JsonHelper.h"
+#include "FactMetaData.h"
+#include "FirmwarePlugin.h"
+#include "FirmwarePluginManager.h"
+#include "QGCApplication.h"
+
+#include <QStandardPaths>
+#include <QJsonDocument>
+#include <QJsonArray>
+
+QGC_LOGGING_CATEGORY(CompInfoGeneralLog, "CompInfoGeneralLog")
+
+const char* CompInfoGeneral::_jsonMetadataTypesKey = "metadataTypes";
+
+CompInfoGeneral::CompInfoGeneral(uint8_t compId, Vehicle* vehicle, QObject* parent)
+    : CompInfo(COMP_METADATA_TYPE_GENERAL, compId, vehicle, parent)
+{
+
+}
+
+void CompInfoGeneral::setUris(CompInfo &compInfo) const
+{
+    const auto& metadataTypeIter = _supportedTypes.constFind(compInfo.type);
+    if (metadataTypeIter == _supportedTypes.constEnd()) {
+        compInfo._uris = {}; // reset
+    } else {
+        compInfo._uris = *metadataTypeIter;
+    }
+}
+
+void CompInfoGeneral::setJson(const QString& metadataJsonFileName, const QString& /*translationJsonFileName*/)
+{
+    if (metadataJsonFileName.isEmpty()) {
+        return;
+    }
+
+    QString         errorString;
+    QJsonDocument   jsonDoc;
+
+    if (!JsonHelper::isJsonFile(metadataJsonFileName, jsonDoc, errorString)) {
+        qCWarning(CompInfoGeneralLog) << "Metadata json file open failed: compid:" << compId << errorString;
+        return;
+    }
+    QJsonObject jsonObj = jsonDoc.object();
+
+    QList<JsonHelper::KeyValidateInfo> keyInfoList = {
+        { JsonHelper::jsonVersionKey,           QJsonValue::Double, true },
+        { _jsonMetadataTypesKey,   QJsonValue::Array,  true },
+    };
+    if (!JsonHelper::validateKeys(jsonObj, keyInfoList, errorString)) {
+        qCWarning(CompInfoGeneralLog) << "Metadata json validation failed: compid:" << compId << errorString;
+        return;
+    }
+
+    int version = jsonObj[JsonHelper::jsonVersionKey].toInt();
+    if (version != 1) {
+        qCWarning(CompInfoGeneralLog) << "Metadata json unsupported version" << version;
+        return;
+    }
+
+    QJsonArray rgSupportedTypes = jsonObj[_jsonMetadataTypesKey].toArray();
+    for (QJsonValue typeValue : rgSupportedTypes) {
+        int type = typeValue["type"].toInt(-1);
+        if (type == -1)
+            continue;
+        Uris uris;
+        uris.uriMetaData = typeValue["uri"].toString();
```

Comment:
I think it would be safer to validate the full type object values as well. I would assume `MetaData` is required and then the rest are optional. But if a uri is specified for any the crc is required.

Reply:
Ok, I added a check for empty uri and crc set